nixos/filesystems: fix special file-systems for systemd-nspawn

Ma27 opened the pull request:
This is a subset of aba55d1b960b4b1817bc2a32deeba447ae51f0a3 (#67336)[1] that I (Ma27) am using for quite a while in my systemd-nspawn setup (without `nixos-container`) to have unprivileged containers.

Recently, @lheckemann  reminded me that this isn't part of upstream NixOS and their setup fails like this when activating config in an nspawn instance (no shared store):

    stderr) activating the configuration...
    stdout) setting up /etc...
    stderr) mount: /dev: permission denied.
    stderr)        dmesg(1) may have more information after failed mount system call.
    stderr) mount: /dev/pts: permission denied.
    stderr)        dmesg(1) may have more information after failed mount system call.
    stderr) mount: /dev/shm: permission denied.
    stderr)        dmesg(1) may have more information after failed mount system call.
    stderr) mount: /run: permission denied.
    stderr)        dmesg(1) may have more information after failed mount system call.
    stdout) Activation script snippet 'specialfs' failed (32)

So I decided to submit this portion again.

[1] Hence I retained the original authorship.


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

uvNikita committed:
nixos/filesystems: fix special file-systems for systemd-nspawn

This is a subset of aba55d1b960b4b1817bc2a32deeba447ae51f0a3 (#67336)[1]
that I (Ma27) am using for quite a while in my systemd-nspawn setup
(without `nixos-container`) to have unprivileged containers.

Recently, Linus reminded me that this isn't part of upstream NixOS and
their setup fails like this when activating config in an nspawn
instance (no shared store):

    stderr) activating the configuration...
    stdout) setting up /etc...
    stderr) mount: /dev: permission denied.
    stderr)        dmesg(1) may have more information after failed mount system call.
    stderr) mount: /dev/pts: permission denied.
    stderr)        dmesg(1) may have more information after failed mount system call.
    stderr) mount: /dev/shm: permission denied.
    stderr)        dmesg(1) may have more information after failed mount system call.
    stderr) mount: /run: permission denied.
    stderr)        dmesg(1) may have more information after failed mount system call.
    stdout) Activation script snippet 'specialfs' failed (32)

So I decided to submit this portion again.

[1] Hence I retained the original authorship.

Co-authored-by: Maximilian Bosch <maximilian@mbosch.me>

aanderse commented:
`isContainer` is far reaching in implications, much further than it was ever intended

people use this value for all sorts of containers, not just `nspawn`, so at bare minimum we should consider this a _potentially_ breaking change...

release notes would be good, but let's see if we can test this change out on a bunch of different container technologies and see if there is any impact?

does `arion` depend on this? what about `incus`? etc...

cc @adamcstephens for informational purposes

adamcstephens reviewed (changes requested):
Thanks @aanderse for the ping. This is indeed a breaking change for incus (and probably lxc) containers.

Perhaps now is a good time to remove some of the overload on `isContainer`? Maybe this PR could instead introduce an option specifically for nspawn, and condition based on that?

adamcstephens commented:
I won't run ofborg on these yet, because they won't succeed but incus can be tested by building `incus-lts.passthru.tests.container-systemd-init` and `incus-lts.passthru.tests.container-legacy-init`

aanderse commented:
> This will be the only place I'll special-case for nspawn for now.

the situation may not be so dire... in many cases we can deal directly with `systemd` units which have a convenient condition directive to check for `nspawn`... so this may be one of the only places we would actually need to use an option like this

arianvp commented:
I actually think we want to get rid of `specialFileSystems` at some point.  All these filesystems are already getting mounted by systemd and now that we have systemd in initrd they also get mounted early enough

aanderse commented:
what about when we don't have `initrd`? looks like stage 2 references `/run` [once](https://github.com/NixOS/nixpkgs/blob/5d4ab3a73eade1160ebd51e4b35aa7a0cc38fe9e/nixos/modules/system/boot/stage-2-init.sh#L121)

that single line should probably be fine to move to a systemd service (or even tmpfiles rule that only runs on boot?)... cool! great suggestion @arianvp

Ma27 commented:
> I actually think we want to get rid of specialFileSystems at some point

Sounds reasonable.
That said, the issue is relatively trivial to fix (even when introducing a `isNspawn`-like option), I think it's still worth fixing.

aanderse commented:
worth fixing, but how? 

so we still need special filesystems or can we already get rid of it and not worry about it anymore? 

🤔

Ma27 commented:
Well, the reason I'm doing this change is to resolve the error I pasted above which occurs when having an nspawn container with its onw Nix store inside and activating a new system configuration in that.

For that problem, there's a fix and just because we should probably remove the entire specialFilesystems part (or even all of non-systemd initrd - which I agree with fwiw), it's still worth fixing problems for as long as it exists.

Ma27 committed:
nixos/containers: add `boot.isNspawnContainer` option

There are a bunch of components such as incus or LXC that also use
`boot.isContainer`, so we'd have to differentiate between "OS container"
and "actually nspawn".

This became necessary for the file-systems part where nspawn takes care
of setting up special filesystems like `/proc`, `/dev` etc., but others
don't.

To allow for a `boot.isContainer` being less overloaded, this introduces
`boot.isNspawnContainer` that is exclusively used for nspawn-specific
things. When `true`, `boot.isContainer = true;` is implied.

Ma27 commented:
Pushed another commit addressing these concerns. WDYT @aanderse @adamcstephens ?

adamcstephens reviewed (approved):
Sorry I dropped the ball on this. Looks good to me.

ElvishJerricco commented on nixos/modules/tasks/filesystems.nix:
```diff
@@ -443,19 +443,19 @@ in
-      "/dev/shm" = { fsType = "tmpfs"; options = [ "nosuid" "nodev" "strictatime" "mode=1777" "size=${config.boot.devShmSize}" ]; };
-      "/dev/pts" = { fsType = "devpts"; options = [ "nosuid" "noexec" "mode=620" "ptmxmode=0666" "gid=${toString config.ids.gids.tty}" ]; };
-
       # To hold secrets that shouldn't be written to disk
       "/run/keys" = { fsType = "ramfs"; options = [ "nosuid" "nodev" "mode=750" ]; };
     } // optionalAttrs (!config.boot.isContainer) {
       # systemd-nspawn populates /sys by itself, and remounting it causes all
       # kinds of weird issues (most noticeably, waiting for host disk device
       # nodes).
       "/sys" = { fsType = "sysfs"; options = [ "nosuid" "noexec" "nodev" ]; };
+    } // optionalAttrs (!config.boot.isNspawnContainer) {
+      "/proc" = { fsType = "proc"; options = [ "nosuid" "noexec" "nodev" ]; };
+      "/run" = { fsType = "tmpfs"; options = [ "nosuid" "nodev" "strictatime" "mode=755" "size=${config.boot.runSize}" ]; };
+      "/dev" = { fsType = "devtmpfs"; options = [ "nosuid" "strictatime" "mode=755" "size=${config.boot.devSize}" ]; };
+      "/dev/shm" = { fsType = "tmpfs"; options = [ "nosuid" "nodev" "strictatime" "mode=1777" "size=${config.boot.devShmSize}" ]; };
+      "/dev/pts" = { fsType = "devpts"; options = [ "nosuid" "noexec" "mode=620" "ptmxmode=0666" "gid=${toString config.ids.gids.tty}" ]; };
```
#301915 was just merged. Might be worth considering setting `enable = false` on these in the appropriate module instead of adding logic here.

ElvishJerricco replied:
Honestly, I think the way you have it here already is fine. That was more of an idea than a goal :P

(we need to kill `specialFileSystems` at some point somehow anyway)

Ma27 committed:
nixos/systemd-machinectl: fix eval with allowAliases=false; unified hierachy is enabled already

Ma27 commented on pkgs/by-name/ni/nixos-container/nixos-container.pl:
```diff
@@ -171,7 +171,7 @@ sub writeNixOSConfig {
 { config, lib, pkgs, ... }:
 
-{ boot.isContainer = true;
+{ boot.isNspawnContainer = true;
```
Good point.
There's no reasonable way to do the upgrade IMHO since you'd have to rewrite the config file.

I extended the release-notes accordingly.

Ma27 committed:
nixos-container: fail if `boot.isNspawnContainer = true` is missing

Inform the poor souls using this in production that
`boot.isNspawnContainer = true;` is required to prevent upgrades with
weird runtime errors after that.

Ma27 commented on pkgs/by-name/ni/nixos-container/package.nix:
```diff
@@ -15,8 +16,9 @@ replaceVarsWith {
-    perl = perl.withPackages (p: [ p.FileSlurp ]);
+    perl = perl.withPackages (p: [ p.FileSlurp p.IPCRun ]);
     su = "${shadow.su}/bin/su";
+    lib = "${path + "/lib"}";
```
I know that this kinda sucks, open for better ideas (I need `versionAtLeast` since Perl's version libraries fail to parse stuff like `25.11pre-git`).

Ma27 commented:
Considering that this has been sitting for a while and even provides an upgrade path for the imperative `nixos-container` use-case, I'm going to merge this in seven days unless any new concerns are being raised.

MatthewCroughan commented:
@Ma27 I think this has broken my setup where I'm setting `nixpkgs = someOldNixpkgs` in a nixos container, as now it cannot evaluate without:

```
       >
       >        error: The option `containers.craterMatthew.boot.isNspawnContainer' does not exist. Definition values:
       >        - In `/nix/store/i5b37k1kib0yfdks2nzqgzksfpbihvp9-source/nixos/modules/virtualisation/nixos-containers.nix': true
```
